sensors/gps/pa6h.c

```diff
@@ -41,8 +52,274 @@ typedef struct {
+{
+	struct termios attr;
+
+	if (tcgetattr(fd, &attr) < 0) {
+		fprintf(stderr, "%s tcgetattr failed\n", PA6H_STR);
+		return -1;
+	}
+
+	/* Backup termios if requested */
+	if (backup != NULL) {
+		*backup = attr;
+	}
+
+	attr.c_iflag = 0;
+	attr.c_oflag = ONLCR;
+	attr.c_cflag = CS8 | CLOCAL;
+	attr.c_lflag = 0;
+
+	/* These parameters are not based upon `picocom` ones */
+	attr.c_cc[VMIN] = 1;
+	attr.c_cc[VTIME] = 0;
+
+	if (cfsetispeed(&attr, baud) < 0 || cfsetospeed(&attr, baud) < 0) {
+		fprintf(stderr, "%s failed to set baudrate", PA6H_STR);
+		return -1;
+	}
+
+	if (tcsetattr(fd, TCSANOW, &attr) < 0) {
+		fprintf(stderr, "%s tcsetattr failed.\n", PA6H_STR);
+		return -1;
+	}
+
+	tcflush(fd, TCIOFLUSH);
+
+	return 0;
+}
+
+
+/*
+* Performs buffered read of gps incoming messages, and saves them to the `inbox`
+*
+* Buffering is inter-read split safe. Message that was only partially read in the first call
+* will be joined with its remainder in the next call (assuming no buffer overflow occurs).
+*/
+static int pa6h_receiver(int fd, pa6h_msg_t *inbox, size_t inboxSz)
+{
+	static char buf[1024] = { 0 };
+	static char *remStart = buf;
+	static int pos = 0;
+	static size_t remLen;
+
+	size_t inboxFill = 0;
+	char *tokenStart = NULL, *tokenEnd = NULL, *nextStartToken, checksum;
+	int i, ret, val, len;
+
+	remLen = &buf[pos] - remStart;
+	pos = 0;
+
+	/* NMEA/PMTK messages are no longer than 128 bytes. Discard partials longer than sizeof(buf)/2 */
+	if (remLen < (sizeof(buf) / 2)) {
+		/* move partial at the end to the beginning of buffer */
+		if (remStart != buf) {
+			memmove(buf, remStart, remLen);
+		}
+		pos = remLen;
+	}
+
+	ret = read(fd, &buf[pos], sizeof(buf) - 1 - pos);
+	remStart = buf;
+	nextStartToken = NULL;
+
+	if (ret >= 0) {
+		/* update position and nul terminate the buffer */
+		pos += ret;
+		buf[pos] = '\0';
+
+		/* find beginning of first message */
+		tokenStart = strchr(buf, '$');
+		tokenEnd = tokenStart;
+		while (tokenStart != NULL) {
+			remStart = tokenStart;
+
+			/* Seek current message`s end. Avoid multiple search for the same 'tokenEnd' if 'tokenStart' is behind it */
+			if (tokenStart >= tokenEnd) {
+				tokenEnd = strchr(tokenStart, '*');
+				if (tokenEnd == NULL) {
+					break;
+				}
+				nextStartToken = strchr(tokenEnd, '$');
+			}
+
+			/* Ensure space for checksum and nul character not interfering with next message. Lack '/r/n' at the end is discarded */
+			len = ((nextStartToken == NULL) ? &buf[pos] : nextStartToken) - tokenEnd;
+			if (len < 3) {
+				break;
+			}
+
+			val = strtol(tokenEnd + 1, (char **)NULL, 16);
+			if (!(val == 0 && errno == EINVAL) && val < 0xff) {
+				len = tokenEnd - tokenStart;
+
+				/* calculate checksum */
+				for (i = 1, checksum = 0; i <= len - 1; i++) {
+					checksum ^= tokenStart[i];
+				}
+
+				/* Save message if checksum matches and there is enough space */
+				if (checksum == val && inboxFill < inboxSz) {
+					inbox[inboxFill].msg = tokenStart;
```

Comment:
If the array `inbox` will be full, you loss valid records.
Here, only check for checksum validity.
```suggestion
				if (checksum == val) {
					inbox[inboxFill].msg = tokenStart;
```

Reply:
Yes, we will loose data if our reads/processing are not as frequent as necessary, Other approach would be to carry _over-inbox_ data to the next read. But this just moves the overload point from PA6H driver to tty driver.